Issue 99

0x1b2c opened the pull request:
- Introduce "unknown" object for gender, birth_year and city.
- Create more groups to improve performance of query "all" genders or "all" birth years.

Close #99

0x1b2c committed:
Use an explicit "unknown" gender for people didn't specify their gender

0x1b2c committed:
Use an explicit UNKNOWN constant for birth_year to indicate people who didn't set birth_year

0x1b2c committed:
Use an explicit "Unknown" city for people didn't specify their city

0x1b2c committed:
Create "missing" groups and remap panel to groups

0x1b2c committed:
Fix Group#to_s

0x1b2c committed:
Extract groups creation code to Group.create_groups!

0x1b2c committed:
Remap people and groups correctly

0x1b2c committed:
Rebulid relations for existing people, groups and panels

0x1b2c committed:
Create panel with all possible condition combinations

0x1b2c commented:
Hey @flyerhzm and @simsicon, would you please help to review this pull request? It changes the under-layer data structure a bit aggressively, it would help a lot if you guys have time to review it, thanks!

0x1b2c committed:
Add progress prompt to very slow people groups relations rebuild

simsicon commented:
Sure, I tried to read it this morning, but did not going well, I will try it tomorrow and comment on some questions.

0x1b2c commented:
Yeah, please don't be hesitant for ask anything. Also it would be easier to understand the point if you review it sequently by commits.

0x1b2c committed:
More progress prompt for the fix data task

0x1b2c committed:
Fix group_spec

0x1b2c committed:
Change people groups relation to one side to avoid MongoDB limitation

0x1b2c committed:
Move Group#add_person to Person#add_to_groups to make it more efficient

0x1b2c committed:
A much more efficent way to rebuild relations for existing data

It should cut down the time to finish this rake task from 100+ hours to
1-2 hours.

0x1b2c committed:
Explicitly drop tweets, stats and rebuild indexes

0x1b2c committed:
Ensure fetch all groups at once to avoid the mystery cursor not found issue

0x1b2c commented:
I've backed up the old database and finished running the migration on server, it actually cost about 3-4 hours.

Please DO NOT deploy develop branch before this pull request be merged, to ensure the database in a consistent state.

0x1b2c committed:
Use more natural #where method

0x1b2c committed:
Refactor DailyStat.record and HourlyStat.record to works more efficient

0x1b2c committed:
More aggressive optimize #record

0x1b2c committed:
Avoid load groups in #record

0x1b2c committed:
Temporarily disable users sampling and tracking

We can't process more data right now.

0x1b2c committed:
Temporarily reduce tweets to process

Since we're reaching the system capacity.

simsicon commented on app/models/person.rb:
```diff
@@ -40,12 +42,21 @@ class Person
+  # already have 2M+ people) in one group will cause the group document reach
+  # that limitation.
+  # http://docs.mongodb.org/manual/reference/object-id/#overview
+  has_and_belongs_to_many :groups, inverse_of: nil, index: true
```
Good to know this, but can we call group.people if there's no has_and_belongs_to many :people in group.rb?

simsicon commented:
I think it's good to ship.
